crates/nu-utils/src/shared_cow.rs

```diff
@@ -0,0 +1,123 @@
+use serde::{Deserialize, Serialize};
+use std::{fmt, ops, sync::Arc};
+
+/// A container that transparently shares a value when possible, but clones on mutate.
+///
+/// Unlike `Arc`, this is only intended to help save memory usage and reduce the amount of effort
+/// required to clone unmodified values with easy to use copy-on-write.
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+#[repr(transparent)]
+pub struct SharedCow<T: Clone>(Arc<T>);
+
+impl<T: Clone> SharedCow<T> {
+    /// Create a new `Shared` value.
+    pub fn new(value: T) -> SharedCow<T> {
+        SharedCow(Arc::new(value))
+    }
+
+    /// Take an exclusive clone of the shared value, or move and take ownership if it wasn't shared.
+    pub fn into_owned(self: SharedCow<T>) -> T {
+        // Optimized: if the Arc is not shared, just unwraps the Arc
+        match Arc::try_unwrap(self.0) {
+            Ok(value) => value,
+            Err(arc) => (*arc).clone(),
+        }
+    }
+
+    /// Convert the `Shared` value into an `Arc`
+    pub fn into_arc(value: SharedCow<T>) -> Arc<T> {
+        value.0
+    }
+
+    /// Return the number of references to the shared value.
+    pub fn ref_count(value: &SharedCow<T>) -> usize {
+        Arc::strong_count(&value.0)
+    }
+}
+
+impl<T> From<T> for SharedCow<T>
+where
+    T: Clone,
+{
+    fn from(value: T) -> Self {
+        SharedCow::new(value)
+    }
+}
+
+impl<T> From<Arc<T>> for SharedCow<T>
+where
+    T: Clone,
+{
+    fn from(value: Arc<T>) -> Self {
+        SharedCow(value)
+    }
+}
+
+impl<T> fmt::Debug for SharedCow<T>
+where
+    T: fmt::Debug + Clone,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // Appears transparent
+        (*self.0).fmt(f)
+    }
+}
+
+impl<T> fmt::Display for SharedCow<T>
+where
+    T: fmt::Display + Clone,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        (*self.0).fmt(f)
+    }
+}
+
+impl<T: Clone> Serialize for SharedCow<T>
+where
+    T: Serialize,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        self.0.serialize(serializer)
+    }
+}
+
+impl<'de, T: Clone> Deserialize<'de> for SharedCow<T>
+where
+    T: Deserialize<'de>,
+{
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        T::deserialize(deserializer).map(Arc::new).map(SharedCow)
+    }
+}
+
+impl<T: Clone> ops::Deref for SharedCow<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<T: Clone> ops::DerefMut for SharedCow<T> {
```

Comment:
Note sure if implementing `DerefMut` is something we want. This could hide a lot of clones, since the compiler will silently insert de-references as necessary.

In that case, I'm not sure if there's a reason to use `SharedCow` as opposed to a regular `Arc`. Maybe for `Serialize` and `Deserialize` implementation?

Reply:
Yes the serde stuff.

I debated that. We could make the mutation explicit with an extra method call, but it will be a whole bunch to change. And on the other hand, this is just a best effort performance enhancement. It should always be at least as good as before.

Reply:
True, it should be as good as before. I am just worried that PRs in the future may use the `DerefMut` without knowing that it can actually perform a clone. It just seems like a potential footgun. 

> but it will be a whole bunch to change.

Yeah... welcome to refactoring in nushell :sweat_smile:. I don't mind pitching in to help.

Also, I am fine with keeping `SharedCow`, since it has the `(De)Serialize` implementations. But I wonder if adding the `rc` feature to `serde` will also work? Not sure if there's any downside to doing this.

Reply:
Adding the `rc` feature came up once before, but decided against it because it would force anyone (including plugins) depending on any of the nushell crates to have `rc` enabled, which can also be a bit of a footgun

I'm not against doing the refactoring if you think it's worth it - we can add `.as_mut()` or perhaps `.into_mut()` instead. That said, doing mutable stuff on it shouldn't lead to a clone most of the time - I can't imagine that it will be hugely surprising for the most part

Reply:
> it would force anyone (including plugins) depending on any of the nushell crates to have `rc` enabled

Oh right, that makes sense. Then yeah, let's keep `SharedCow` for this.

> I can't imagine that it will be hugely surprising for the most part

Well, I think Rust code tends to be explicit where possible. Potentially cloning on deref seems like very unexpected behavior. The Rust [docs](https://doc.rust-lang.org/std/ops/trait.Deref.html#when-to-implement-deref-or-derefmut) for `Deref` specifically mention that:
2. the implementation of the deref function should be cheap; and
3. users of the type will not be surprised by any deref coercion behaviour.

>  is someone really likely to rethink whether they need a mutable ref just because they have to call `.into_mut()`

Maybe, but it at least gives one place where we can  document that mutable references can potentially clone.

> we can add `.as_mut()` or perhaps `.into_mut()` instead

To be a little pedantic :stuck_out_tongue: , this would add an inherent method on a generic type that implements `Deref`. So, either we would have to drop the `Deref` impl, or use a non-inherent method. (I think you're already aware of this, since the other methods you added take `SharedCow<T>` instead of `self`.)

Reply:
Ok, will remove the `DerefMut`.

> To be a little pedantic 😛 , this would add an inherent method on a generic type that implements Deref. So, either we would have to drop the Deref impl, or use a non-inherent method. (I think you're already aware of this, since the other methods you added take SharedCow<T> instead of self.)

True, but also not totally without precedent. If I were to follow the `Cow` API as the name implies, I think it should actually be `to_mut(&mut self)`: https://doc.rust-lang.org/stable/std/borrow/enum.Cow.html#method.to_mut

Reply:
Oh true, ok that sound good.